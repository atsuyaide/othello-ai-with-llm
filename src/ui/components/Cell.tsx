import * as style from "@app/ui/constants/style";
import { CellState, Color, Place } from "@app/ui/types";
import { CSSProperties } from "react";
import * as Constants from "../constants/style";

export interface CellProps {
  place: Place;
  state: CellState;
  scale?: number;
  highlight: boolean;
  onClick: () => void;
}

export const Cell = (props: CellProps) => (
  <div
    style={cellStyle(props.highlight)}
    onClick={props.onClick}
    data-testid="cell"
  >
    {props.state === "." && null}
    {props.state === "b" && (
      <span data-testid="stone" style={stoneStyle(props.state)} />
    )}
    {props.state === "w" && (
      <span data-testid="stone" style={stoneStyle(props.state)} />
    )}
    {props.state === "*" && (
      <span data-testid="marker" style={movableStyle()} />
    )}
  </div>
);

const cellStyle = (highlight: boolean, scale: number = 1): CSSProperties => ({
  float: "left",
  position: "relative",
  width: style.cellWidth * scale,
  height: style.cellWidth * scale,
  margin: style.cellMargin * scale,
  background: highlight ? Constants.boardColorHighlight : Constants.boardColor,
  alignItems: "center",
});

const stoneStyle = (color: Color, scale: number = 1) => ({
  display: "inline-block",
  width: (style.cellWidth - style.stoneMargin * 2) * scale,
  height: (style.cellWidth - style.stoneMargin * 2) * scale,
  borderRadius: style.cellWidth * scale,
  background:
    color == "b" ? Constants.stoneColorBlack : Constants.stoneColorWhite,
  margin: style.stoneMargin,
});

const movableStyle = (scale: number = 1) => ({
  display: "inline-block",
  width: (style.cellWidth - style.movableMargin * 2) * scale,
  height: (style.cellWidth - style.movableMargin * 2) * scale,
  borderRadius: style.cellWidth * scale,
  background: Constants.movableCellColor,
  margin: style.movableMargin,
});
