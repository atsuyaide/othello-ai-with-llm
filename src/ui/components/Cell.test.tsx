import { fireEvent, render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { Cell } from "./Cell";

describe("Cell", () => {
  it("renders correctly", () => {
    const onClickMock = vi.fn();
    const { container } = render(
      <Cell
        place={{ x: 1, y: 2 }}
        state="b"
        highlight={false}
        onClick={onClickMock}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("calls onClick function when clicked", () => {
    const onClickMock = vi.fn();
    const { container } = render(
      <Cell
        place={{ x: 1, y: 2 }}
        state="w"
        highlight={true}
        onClick={onClickMock}
      />
    );

    expect(container.firstChild).not.toBeNull();
    const firstChild = container.firstChild as Element;
    fireEvent.click(firstChild);
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  it("renders null when state is '.'", () => {
    const onClickMock = vi.fn();
    const { container } = render(
      <Cell
        place={{ x: 1, y: 2 }}
        state="."
        highlight={false}
        onClick={onClickMock}
      />
    );

    expect(container.firstChild).toBeNull();
  });

  it("renders black stone when state is 'b'", () => {
    const onClickMock = vi.fn();
    const { container } = render(
      <Cell
        place={{ x: 1, y: 2 }}
        state="b"
        highlight={false}
        onClick={onClickMock}
      />
    );

    expect(container.firstChild).toContainHTML(
      '<span style="color: black;"></span>'
    );
  });

  // it("renders white stone when state is 'w'", () => {
  //   const onClickMock = vi.fn();
  //   const { container } = render(
  //     <Cell
  //       place={{ x: 1, y: 2 }}
  //       state="w"
  //       highlight={false}
  //       onClick={onClickMock}
  //     />
  //   );

  //   expect(container.firstChild).toContainHTML(
  //     '<span style="color: white;"></span>'
  //   );
  // });

  // it("renders movable style when state is '*'", () => {
  //   const onClickMock = vi.fn();
  //   const { container } = render(
  //     <Cell
  //       place={{ x: 1, y: 2 }}
  //       state="*"
  //       highlight={false}
  //       onClick={onClickMock}
  //     />
  //   );

  //   expect(container.firstChild).toContainHTML(
  //     '<span style="border: 2px solid red;"></span>'
  //   );
  // });
});
