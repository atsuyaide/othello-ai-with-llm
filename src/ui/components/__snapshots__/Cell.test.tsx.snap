// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cell > renders correctly 1`] = `
<div
  style="float: left; width: 40px; height: 40px; margin: 1px; background: rgb(0, 153, 0);"
>
  <span
    data-testid="stone"
    style="display: inline-block; width: 36px; height: 36px; border-radius: 40px; background: rgb(32, 39, 32); margin: 2px;"
  />
</div>
`;
